File: Emition of sequence variants and inversions/driver/HW7.java
package abhamare_hw7EC.driver;

import abhamare_hw7EC.invert.Inverter;
import abhamare_hw7EC.sequence.Extractor;
import abhamare_hw7EC.sequence.Repeater;
import abhamare_hw7EC.sequence.Sequence;
import abhamare_hw7EC.sequence.Variator;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.Random;
import java.util.Scanner;


import static abhamare_hw7EC.sequence.State.ACTIVE;

public class HW7
{
    static Random random = new Random();
    static Scanner keyboard = new Scanner(System.in);
    static int totalWords = 0;

    public static void main(String[] args)
    {
        String line = "";
        int SIZE = 9;
        Sequence[] sequence = new Sequence[SIZE];
        String[] words;
        String answer;
        String input;
        char repeat = 'y';

        Scanner keyboard = new Scanner(System.in);

        System.out.println("\nWelcome to the Sequence tester!");
        words = readFile();

        int count = 3;
        for (int k = 0; k < count; k++)
        {
            int index = random.nextInt(totalWords - 1);
            System.out.println("\n----------");

            System.out.println("\nTesting word " + (k + 1) + ":");

            Repeater repeater = new Repeater(words[index], ACTIVE);

            for(int l = 0; l < 3; l++)
            {
                sequence[l] = repeater;
                String subSequence = words[index].substring(0, words[index].length() / 2);
                Extractor extractor = new Extractor(words[index], subSequence, ACTIVE);

                sequence[l + 1] = extractor;
                Variator variator = new Variator(words[index], ACTIVE);
                sequence[l + 2] = variator;
                System.out.println("\nemit: " + sequence[l].emitWord().trim());
                Inverter inverter = new Inverter(sequence[l]);
                System.out.println("invert: " + inverter.invertElements().trim());

            }

            while(repeat == 'y' || repeat == 'Y')
            {
                System.out.print("\nWhat is your guess? ");
                answer = keyboard.nextLine();

                if (answer.trim().equals(words[index]))
                {
                    System.out.println("That's correct!");
                    break;
                }
                else
                {
                    System.out.println("That is not correct.");
                    System.out.print("Guess again (y/n)? ");
                    input = keyboard.next();   // Read a line.
                    keyboard.nextLine();
                    repeat = input.charAt(0);
                }
            }
        }

        System.out.println("\n----------");

        testAdditionalFunctionality(words, line);

        System.out.println("\n----------");
        System.out.println("\nThanks for using the Sequence tester! \n");

    }

    public static String[] readFile()
    {
        BufferedReader fileReader;
        int numberOfLines = 0;
        String[] lines = new String[50];
        try {

            fileReader = new BufferedReader(new FileReader
                    ("/Users/rishi/Desktop/CPSC5011/abhamare__hw7EC/data/words.dat"));
            while (fileReader.readLine() != null)
            {
                lines[numberOfLines] = fileReader.readLine();
                numberOfLines++;
            }
        } catch (FileNotFoundException e)
        {
            e.printStackTrace();
        } catch (IOException e)
        {
            e.printStackTrace();
        }
        totalWords = numberOfLines;
        return lines;
    }


public static void testAdditionalFunctionality(String[] words, String line)
    {
        String input;
        char repeat = 'y';

        System.out.println("Testing word 4 (additional functionality): \n");
        do
        {

            int index = random.nextInt(totalWords - 1);

            Repeater repeater = new Repeater(words[index], ACTIVE);

            if (repeater.getWord().isEmpty())
            {
                System.out.println("A Word " + line + " length is less than 3 characters long ");

                repeat = 'y';
                continue;
            }

            Inverter inverter = new Inverter(repeater);
            System.out.println("Invert: " + inverter.invertElements().trim());

            do
            {
                //int endIndex = random.nextInt() % (line.length() - 2) + 1;
                int endIndex = random.nextInt(50) % (words[index].length() / 2) + 1;
                //String subSequence = words[index].substring(0, (words[index].length() / 2) + 1);
                String subSequence = words[index].substring(0, endIndex);

                System.out.println("\nprovide a subsequence: " + subSequence);

                Extractor extractor = new Extractor(words[index], subSequence, ACTIVE);

                System.out.println("emit: " + extractor.emitWord().trim());

                System.out.print("\nProvide another subsequence (y/n)? ");
                input = keyboard.next();  // Read a line.
                repeat = input.charAt(0);

            } while (repeat == 'y' || repeat == 'Y');

            do {
                String response;
                System.out.print("\nWhat is your guess? ");
                response = keyboard.next();

                if (response.trim().equals(words[index]))
                {
                    System.out.println("That's correct!");
                    break;
                }
                else
                {
                    System.out.println("That is not correct.");
                    System.out.print("Guess again (y/n)? ");
                    input = keyboard.next();  // Read a line.
                    repeat = input.charAt(0);

                }
            } while (repeat == 'y' || repeat == 'Y');

            System.out.print("\nWould you like to test another word (y/n)? ");
            input = keyboard.next();  // Read a line.
            repeat = input.charAt(0);

        } while (repeat == 'y' || repeat == 'Y');

    }
}

File: Emition of sequence variants and inversions EC/hw7.cpp
/**
 * Author: Ashlesha Bhamare
 * File name: hw7.cpp
 * Version: 1.0
 * References: none
 * Description: A class that tests all the functionality of other classes
 * Dependencies: A main class which requires all header files for accessing
 * their functionality.
 * Emit behaviour of Sequence class is tested with invert method of Inverter
 * class
 * (List any additional dependencies. Are all behaviors and
 * attributes inherited? Is anything suppressed?)
 * State: Any state changes observed in the DerivedClass.
 * (should align with the pre/post conditions)
 * Assumptions:
 * - Assumption 1 for DerivedClass
 * - Assumption 2 for DerivedClass
*/

#include <iostream>
#include <fstream>

#include "Sequence.h"
#include "Repeater.h"
#include "Inverter.h"
#include "Variator.h"
#include "Extractor.h"

using namespace std;

static bool isSeeded = false;

// This is the fileReading function which reads random lines from the words.dat
// file
// Precondition: The file should not be empty
// Postconditon: The words array will be filled with random line word.
string fileReading();

// This is testAdditionalFunctionality function which tests additional
// functionality like provides subsequence
// Precondition: The lines of words should be pass
// Postconditon: provides subsequences and emits the word.
void testAdditionalFunctionality(string);

// This is the mainn function
// Precondition: The state should be ready to set
// Postconditon: The state is either set to Active or Inactive.
int main()
{
    string line;                                 // To hold line
    const int SIZE = 9;                          // To hold size
    Sequence *sequenceArray[SIZE];               // To hold sequenceArray
    const int wordLength = 3;                    // To hold wordLength
    string word[wordLength];                     // To hold array of word

    cout << "\nWelcome to the Sequence tester! " << endl;


    for (int i = 0; i < wordLength; i++)
    {
        line = fileReading();
        word[i] = line;
    }

    int count = 3;
    for (int i = 0; i < count; i++)
    {
        int j = 0;
        if (i > 0)
        {
            j = i * 3;
        }
        line = word[i];

        Repeater *repeater = new Repeater(line, ACTIVE);

        if (repeater ->getWord().empty())
        {
            cout << "A Word " << line << " length is less than 3 characters "
                                         "long ";
            count = count - 1;
            continue;
        }

        sequenceArray[j] = repeater;

        string subSequence = line.substr(0, line.length()/2);

        Extractor *extractor = new Extractor(line, subSequence, ACTIVE);
        sequenceArray[j + 1] = extractor;


        Variator *variator = new Variator(line, ACTIVE);
        sequenceArray[j + 2] = variator;

    }

    int i = 0;
    const int WORDS = 3;

    for (int j = 0 ; j < WORDS; j++) {
        cout << "\n----------" << endl;

        cout << "Testing word " << (j + 1) << ": " << endl;
        line = word[j];

        if (j > 0)
        {
            i = j * 3;
        }
        int limit = 0;
        // checks for limit and size
        while (limit < 3 && i < SIZE)
        {
            if (limit == 3)
            {
                break;
            }
            // emits the word
            cout << "\nemit: " << sequenceArray[i] ->emitWord() << endl;
            Inverter inverter(sequenceArray[i]);
            // Invert the word
            cout << "invert: " << inverter.invertElements() << endl;
            limit++;
            i++;
        }
        string answer;
        string response = "y";
        // checking for response
        while(response == "y" || response == "Y")
        {
            cout << "\nWhat is your guess? " ;
            cin >> answer;

            if (answer.compare(line) == 0)
            {
                cout << "That's correct!" << endl;
                break;
            }
            else
            {
                cout << "That is not correct." << endl;
                cout << "Guess again (y/n)? ";
                cin >> response;
            }
        }
    }

    cout << "\n----------" << endl;
    // calls testAdditionalFunctionality function
    testAdditionalFunctionality(line);

    cout << "\n----------" << endl;
    cout << "\nThanks for using the Sequence tester! \n" << endl;

    return 0;
}

string fileReading()
{
    int numberOfLines = 0;                        // To hold numberOfLines
    int randomNumber;                             // To hold randomNumber
    string line;                                  // to hold line
    ifstream inFile;

    // To open file
    inFile.open("words.dat");

    if (!inFile)
    {
        cout << "This file does not exists. Try again";
        exit(1);           // terminate with error
    }
    else
    {
        while (getline(inFile, line))
        {
            numberOfLines++;
        }

        if (!isSeeded)
        {
            srand(time(NULL));
            isSeeded = true;
        }

        randomNumber = rand() % numberOfLines + 1;
        int currentLine = 0;
        ifstream readFile;

        readFile.open("words.dat");
        while (getline(readFile, line))
        {
            if (currentLine == randomNumber)
            {
                break;
            }
            currentLine++;
        }
        inFile.close();                           // To close file
        readFile.close();                         // To close file
    }
    return line;
}

void testAdditionalFunctionality(string line)
{
    string input;
    string response;

    cout << "Testing word 4 (additional functionality): " << endl;
    cout << endl;

    do {
        response = "n";

        line = fileReading();

        Repeater repeater(line, ACTIVE);

        if (repeater.getWord().empty())
        {
            cout << "A Word " << line << " length is less than 3 characters "
                                         "long ";
            response = 'y';
            continue;
        }

        Inverter inverter(&repeater);
        cout << "Invert: " << inverter.invertElements() << endl;

        do {
            if (!isSeeded)
            {
                srand(time(NULL));
                isSeeded = true;
            }

            int endIndex = rand() % (line.length() - 2) + 1;

            string subSequence = line.substr(0, endIndex);
            cout << "\nprovide a subsequence: " << subSequence << endl;

            Extractor extractor(line, subSequence, ACTIVE);
            cout << "emit: " << extractor.emitWord() << endl;

            cout << "\nProvide another subsequence (y/n)? ";
            cin >> input;

        } while (input == "y" || input == "Y");

        do {
           cout << "\nWhat is your guess? ";
           cin >> response;

           if (response == line)
           {
               cout << "That's correct!" << endl;
           }
           else
           {
               cout << "That is not correct." << endl;
               cout << "Guess again (y/n)? ";
               cin >> response;
           }
        } while (response == "y" || response == "Y");

        cout << "\nWould you like to test another word (y/n)? ";
        cin >> response;

    } while (response == "y" || response == "Y");
}